File: SDK/CentralService/python/service/erp_centralservice_service/transport.py
"""CentralService Python service SDK 的多端点传输实现。"""

from __future__ import annotations

import json
import os
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Optional

DEFAULT_BASE_URL = "http://127.0.0.1:5000"
DEFAULT_TIMEOUT_SEC = 5.0
DEFAULT_MAX_ATTEMPTS = 2


@dataclass
class CentralServiceError(Exception):
    """用于 HTTP 与业务层失败的标准化 SDK 异常。"""

    http_status: int
    method: str
    url: str
    kind: str
    message: str
    error_code: Optional[int] = None
    raw_body: str = ""

    def __str__(self) -> str:  # pragma: no cover
        return f"{self.kind} HTTP {self.http_status} {self.method} {self.url}: {self.message}"


@dataclass
class CentralServiceCircuitBreakerOptions:
    """单端点熔断器配置。"""

    failure_threshold: int
    break_duration_minutes: int
    recovery_threshold: int


@dataclass
class CentralServiceEndpointOptions:
    """单端点配置。"""

    base_url: str
    priority: int = 0
    max_attempts: Optional[int] = None
    circuit_breaker: Optional[CentralServiceCircuitBreakerOptions] = None
    order: int = 0


class CircuitBreakerState:
    """在进程内维护单个端点的熔断状态。"""

    def __init__(self, options: CentralServiceCircuitBreakerOptions) -> None:
        self.failure_threshold = max(1, int(options.failure_threshold))
        self.break_duration_sec = max(1, int(options.break_duration_minutes)) * 60
        self.recovery_threshold = max(1, int(options.recovery_threshold))
        self.mode = "closed"
        self.failure_count = 0
        self.half_open_success_count = 0
        self.open_until = 0.0

    def try_allow_request(self, now: float) -> tuple[bool, Optional[str]]:
        if self.mode == "open":
            if now >= self.open_until:
                self.mode = "half-open"
                self.failure_count = 0
                self.half_open_success_count = 0
                return True, None

            remaining = max(1, int(self.open_until - now + 0.999))
            return False, f"熔断开启，剩余约 {remaining} 秒"

        return True, None

    def report_success(self) -> None:
        if self.mode == "half-open":
            self.half_open_success_count += 1
            if self.half_open_success_count >= self.recovery_threshold:
                self.mode = "closed"
                self.failure_count = 0
                self.half_open_success_count = 0
                self.open_until = 0.0
            return

        self.failure_count = 0

    def report_failure(self, now: float) -> bool:
        if self.mode == "half-open":
            self._open(now)
            return False

        self.failure_count += 1
        if self.failure_count >= self.failure_threshold:
            self._open(now)
            return False
        return True

    def _open(self, now: float) -> None:
        self.mode = "open"
        self.failure_count = 0
        self.half_open_success_count = 0
        self.open_until = now + self.break_duration_sec


@dataclass
class TransportEndpoint:
    """运行时归一化后的端点定义。"""

    base_url: str
    priority: int
    max_attempts: int
    order: int
    circuit_breaker: Optional[CircuitBreakerState]


@dataclass
class TransportResult:
    """一次传输成功后的原始 HTTP 结果。"""

    base_url: str
    url: str
    attempt: int
    max_attempts: int
    status_code: int
    body_text: str
    skipped_endpoints: List[str]


class TransportExhaustedError(Exception):
    """所有候选端点都因传输异常或熔断跳过而无法完成请求。"""

    def __init__(
        self,
        method: str,
        path: str,
        last_url: Optional[str],
        skipped_endpoints: List[str],
        failure_summaries: List[str],
    ) -> None:
        segments: List[str] = []
        if skipped_endpoints:
            segments.append("跳过端点: " + "; ".join(skipped_endpoints))
        if failure_summaries:
            segments.append("失败详情: " + "; ".join(failure_summaries))
        if not segments:
            segments.append("未找到可用的中心服务端点。")

        self.method = method
        self.path = path
        self.last_url = last_url or ""
        self.raw_detail = " | ".join(segments)
        super().__init__("中心服务调用失败，所有可用端点均已耗尽。 " + self.raw_detail)


def normalize_base_url(base_url: str) -> str:
    """去掉 BaseUrl 尾部斜杠。"""

    if not base_url:
        raise ValueError("base_url is required")
    return base_url.rstrip("/")


def normalize_max_attempts(value: Optional[int]) -> int:
    """归一化单端点最大尝试次数。"""

    try:
        normalized = int(value if value is not None and value != "" else DEFAULT_MAX_ATTEMPTS)
    except (TypeError, ValueError):
        normalized = DEFAULT_MAX_ATTEMPTS
    return normalized if normalized >= 1 else DEFAULT_MAX_ATTEMPTS


def coerce_int(value: Any, default: int) -> int:
    """将任意值转换为整数。"""

    try:
        return int(value if value is not None and value != "" else default)
    except (TypeError, ValueError):
        return default


def normalize_circuit_breaker(raw: Optional[Dict[str, Any]]) -> Optional[CentralServiceCircuitBreakerOptions]:
    """归一化熔断配置。"""

    if raw is None:
        return None

    return CentralServiceCircuitBreakerOptions(
        failure_threshold=max(1, coerce_int(raw.get("failureThreshold", raw.get("failure_threshold", 1)), 1)),
        break_duration_minutes=max(1, coerce_int(raw.get("breakDurationMinutes", raw.get("break_duration_minutes", 1)), 1)),
        recovery_threshold=max(1, coerce_int(raw.get("recoveryThreshold", raw.get("recovery_threshold", 1)), 1)),
    )


def normalize_endpoints(base_url: Optional[str], endpoints: Optional[Iterable[Dict[str, Any]]]) -> List[TransportEndpoint]:
    """将原始输入转换为排序后的端点数组。"""

    raw_endpoints = list(endpoints or [])
    if not raw_endpoints:
        raw_endpoints = [{"baseUrl": base_url}]

    normalized: List[TransportEndpoint] = []
    for order, raw_endpoint in enumerate(raw_endpoints):
        if not raw_endpoint:
            continue
        endpoint_base_url = raw_endpoint.get("baseUrl") or raw_endpoint.get("base_url")
        if not endpoint_base_url:
            continue
        breaker_options = normalize_circuit_breaker(raw_endpoint.get("circuitBreaker") or raw_endpoint.get("circuit_breaker"))
        normalized.append(
            TransportEndpoint(
                base_url=normalize_base_url(str(endpoint_base_url)),
                priority=coerce_int(raw_endpoint.get("priority", 0), 0),
                max_attempts=normalize_max_attempts(raw_endpoint.get("maxAttempts", raw_endpoint.get("max_attempts"))),
                order=order,
                circuit_breaker=CircuitBreakerState(breaker_options) if breaker_options else None,
            )
        )

    normalized.sort(key=lambda item: (item.priority, item.order))
    if not normalized:
        raise ValueError("at least one central service endpoint is required")
    return normalized


def load_sdk_options_from_env(timeout_sec: Optional[float] = None) -> Dict[str, Any]:
    """从环境变量读取 SDK 选项。"""

    resolved_timeout_sec = timeout_sec
    if resolved_timeout_sec is None:
        raw_timeout_ms = os.environ.get("CENTRAL_SERVICE_TIMEOUT_MS")
        if raw_timeout_ms:
            try:
                resolved_timeout_sec = max(1, int(raw_timeout_ms)) / 1000.0
            except ValueError:
                resolved_timeout_sec = DEFAULT_TIMEOUT_SEC
        else:
            resolved_timeout_sec = DEFAULT_TIMEOUT_SEC

    raw_endpoints = os.environ.get("CENTRAL_SERVICE_ENDPOINTS_JSON")
    if raw_endpoints:
        return {
            "endpoints": json.loads(raw_endpoints),
            "timeout_sec": resolved_timeout_sec,
        }

    return {
        "base_url": os.environ.get("CENTRAL_SERVICE_BASEURL", DEFAULT_BASE_URL).rstrip("/"),
        "timeout_sec": resolved_timeout_sec,
    }


def build_url(base_url: str, path: str) -> str:
    """拼接 BaseUrl 与路径。"""

    base = normalize_base_url(base_url)
    normalized_path = path if path.startswith("/") else f"/{path}"
    return f"{base}{normalized_path}"


def looks_like_json(text: str) -> bool:
    """判断响应体是否看起来像 JSON。"""

    if not text:
        return False
    trimmed = text.lstrip()
    return trimmed.startswith("{") or trimmed.startswith("[")


def parse_error(method: str, url: str, status: int, body_text: str) -> CentralServiceError:
    """将错误响应转换为标准化异常。"""

    raw_body = body_text or ""
    trimmed = raw_body.strip()
    if not looks_like_json(trimmed):
        return CentralServiceError(status, method, url, "PlainText", trimmed or f"HTTP {status}", raw_body=raw_body)

    try:
        parsed = json.loads(trimmed)
    except Exception:
        return CentralServiceError(status, method, url, "PlainText", trimmed or f"HTTP {status}", raw_body=raw_body)

    if isinstance(parsed, dict) and "errors" in parsed:
        return CentralServiceError(status, method, url, "ValidationProblemDetails", str(parsed.get("title") or "Validation error"), raw_body=raw_body)
    if isinstance(parsed, dict) and parsed.get("title") and parsed.get("status") is not None:
        return CentralServiceError(status, method, url, "ProblemDetails", str(parsed.get("title") or "ProblemDetails"), raw_body=raw_body)
    if isinstance(parsed, dict) and "success" in parsed:
        return CentralServiceError(
            status,
            method,
            url,
            "ApiResponse",
            str(parsed.get("errorMessage") or "ApiResponse error"),
            error_code=parsed.get("errorCode"),
            raw_body=raw_body,
        )

    return CentralServiceError(status, method, url, "Unknown", "Unknown error", raw_body=raw_body)


def append_transport_context(message: str, transport: TransportResult) -> str:
    """附加多端点上下文。"""

    segments = [
        f"端点={transport.base_url}",
        f"尝试={transport.attempt}/{transport.max_attempts}",
    ]
    if transport.skipped_endpoints:
        segments.append("已跳过=" + "、".join(transport.skipped_endpoints))
    return f"{message or ''} ({'; '.join(segments)})"


def create_parsed_error(method: str, transport: TransportResult, error: CentralServiceError) -> CentralServiceError:
    """基于传输结果补全错误上下文。"""

    return CentralServiceError(
        http_status=error.http_status,
        method=method,
        url=transport.url,
        kind=error.kind,
        message=append_transport_context(error.message, transport),
        error_code=error.error_code,
        raw_body=error.raw_body,
    )


def create_transport_error(method: str, error: TransportExhaustedError) -> CentralServiceError:
    """将端点耗尽转换为对外异常。"""

    return CentralServiceError(
        http_status=503,
        method=method,
        url=error.last_url or error.path,
        kind="Transport",
        message=str(error),
        raw_body=error.raw_detail,
    )


def http_json(method: str, url: str, body: Optional[Dict[str, Any]], timeout_sec: float) -> tuple[int, str]:
    """发送 JSON 请求并返回原始响应。"""

    data = None if body is None else json.dumps(body).encode("utf-8")
    req = urllib.request.Request(url=url, method=method)
    req.add_header("Accept", "application/json")
    if data is not None:
        req.add_header("Content-Type", "application/json; charset=utf-8")
        req.data = data

    try:
        with urllib.request.urlopen(req, timeout=timeout_sec) as resp:
            status = int(getattr(resp, "status", 0) or 0)
            text = resp.read().decode("utf-8")
            return status, text
    except urllib.error.HTTPError as exc:
        try:
            text = exc.read().decode("utf-8")
        except Exception:
            text = ""
        return int(exc.code), text


class MultiEndpointTransport:
    """按优先级处理多端点请求。"""

    def __init__(self, endpoints: List[TransportEndpoint], timeout_sec: float) -> None:
        self.endpoints = endpoints
        self.timeout_sec = timeout_sec

    def send(self, method: str, path: str, body: Optional[Dict[str, Any]]) -> TransportResult:
        skipped_endpoints: List[str] = []
        failure_summaries: List[str] = []
        last_url: Optional[str] = None

        for endpoint in self.endpoints:
            now = time.time()
            if endpoint.circuit_breaker is not None:
                allowed, reason = endpoint.circuit_breaker.try_allow_request(now)
                if not allowed:
                    skipped_endpoints.append(f"{endpoint.base_url}（{reason}）")
                    continue

            for attempt in range(1, endpoint.max_attempts + 1):
                url = build_url(endpoint.base_url, path)
                last_url = url
                try:
                    status_code, body_text = http_json(method, url, body, self.timeout_sec)
                    if endpoint.circuit_breaker is not None:
                        endpoint.circuit_breaker.report_success()
                    return TransportResult(
                        base_url=endpoint.base_url,
                        url=url,
                        attempt=attempt,
                        max_attempts=endpoint.max_attempts,
                        status_code=status_code,
                        body_text=body_text,
                        skipped_endpoints=list(skipped_endpoints),
                    )
                except (urllib.error.URLError, TimeoutError, OSError, ValueError) as exc:
                    should_retry = True
                    if endpoint.circuit_breaker is not None:
                        should_retry = endpoint.circuit_breaker.report_failure(time.time())
                    failure_summaries.append(
                        f"{endpoint.base_url} 第 {attempt}/{endpoint.max_attempts} 次失败：{type(exc).__name__}: {exc}"
                    )
                    if not should_retry:
                        break

        raise TransportExhaustedError(method, path, last_url, skipped_endpoints, failure_summaries)

File: SDK/CentralService/rust/client/src/transport.rs
use crate::error::{Result, SdkError};
use crate::http::HttpClient;
use crate::options::{CentralServiceEndpointOptions, DiscoveryClientOptions};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

/// 多中心端点传输层的原始 HTTP 结果。
#[derive(Debug, Clone)]
pub struct TransportResponse {
    /// 实际命中的端点根地址。
    pub base_url: String,
    /// 实际请求 URL。
    pub url: String,
    /// 当前端点第几次尝试。
    pub attempt: usize,
    /// 当前端点允许的最大尝试次数。
    pub max_attempts: usize,
    /// HTTP 状态码。
    pub status: u16,
    /// 原始响应体。
    pub body: Vec<u8>,
    /// 本次请求前已被熔断跳过的端点摘要。
    pub skipped_endpoints: Vec<String>,
}

#[derive(Debug, Clone)]
pub struct MultiEndpointTransport {
    endpoints: Vec<TransportEndpoint>,
    timeout: Duration,
}

#[derive(Debug, Clone)]
struct TransportEndpoint {
    base_url: String,
    max_attempts: usize,
    circuit_breaker: Option<Arc<Mutex<CircuitBreakerState>>>,
}

#[derive(Debug)]
struct CircuitBreakerState {
    failure_threshold: usize,
    break_duration: Duration,
    recovery_threshold: usize,
    mode: CircuitBreakerMode,
    failure_count: usize,
    half_open_success_count: usize,
    open_until: Option<Instant>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum CircuitBreakerMode {
    Closed,
    Open,
    HalfOpen,
}

impl MultiEndpointTransport {
    /// 基于顶层配置创建多端点传输层。
    pub fn new(options: &DiscoveryClientOptions) -> Self {
        let endpoints = options
            .endpoints
            .iter()
            .map(TransportEndpoint::from_options)
            .collect::<Vec<_>>();

        Self {
            endpoints,
            timeout: options.timeout,
        }
    }

    /// 发送一次带端点切换能力的请求。
    pub fn send(
        &self,
        method: &str,
        path_and_query: &str,
        body: Option<&[u8]>,
        content_type: Option<&str>,
    ) -> Result<TransportResponse> {
        let mut skipped_endpoints = Vec::new();
        let mut failures = Vec::new();
        let mut last_url = String::new();

        for endpoint in &self.endpoints {
            if let Some(reason) = endpoint.circuit_breaker_skip_reason() {
                skipped_endpoints.push(format!("{}（{}）", endpoint.base_url, reason));
                continue;
            }

            for attempt in 1..=endpoint.max_attempts {
                let client = HttpClient::new(&endpoint.base_url)?.with_timeout(self.timeout);
                let url = build_full_url(&endpoint.base_url, path_and_query);
                last_url = url.clone();

                match client.request(method, path_and_query, body, content_type) {
                    Ok(response) => {
                        endpoint.report_success();
                        return Ok(TransportResponse {
                            base_url: endpoint.base_url.clone(),
                            url,
                            attempt,
                            max_attempts: endpoint.max_attempts,
                            status: response.status,
                            body: response.body,
                            skipped_endpoints,
                        });
                    }
                    Err(SdkError::Io(error)) => {
                        endpoint.report_failure();
                        failures.push(format!(
                            "{} 第 {attempt}/{} 次失败：{}",
                            endpoint.base_url, endpoint.max_attempts, error
                        ));
                    }
                    Err(other) => return Err(other),
                }
            }
        }

        let mut segments = Vec::new();
        if !skipped_endpoints.is_empty() {
            segments.push(format!(
                "跳过端点: {}",
                skipped_endpoints.join("; ")
            ));
        }
        if !failures.is_empty() {
            segments.push(format!("失败详情: {}", failures.join("; ")));
        }
        if segments.is_empty() {
            segments.push("未找到可用的中心服务端点".to_string());
        }

        Err(SdkError::Transport {
            message: format!(
                "中心服务调用失败，所有可用端点均已耗尽。 {}",
                segments.join(" | ")
            ),
            detail: format!("method={method}; path={path_and_query}; lastUrl={last_url}"),
        })
    }
}

impl TransportEndpoint {
    fn from_options(options: &CentralServiceEndpointOptions) -> Self {
        Self {
            base_url: options.base_url.clone(),
            max_attempts: options.normalized_max_attempts(),
            circuit_breaker: options.circuit_breaker.as_ref().map(|breaker| {
                Arc::new(Mutex::new(CircuitBreakerState::new(
                    breaker.failure_threshold,
                    Duration::from_secs(breaker.break_duration_minutes * 60),
                    breaker.recovery_threshold,
                )))
            }),
        }
    }

    fn circuit_breaker_skip_reason(&self) -> Option<String> {
        let state = self.circuit_breaker.as_ref()?;
        let mut state = state.lock().expect("circuit breaker lock poisoned");
        state.try_allow_request()
    }

    fn report_success(&self) {
        if let Some(state) = &self.circuit_breaker {
            state
                .lock()
                .expect("circuit breaker lock poisoned")
                .report_success();
        }
    }

    fn report_failure(&self) {
        if let Some(state) = &self.circuit_breaker {
            state
                .lock()
                .expect("circuit breaker lock poisoned")
                .report_failure();
        }
    }
}

impl CircuitBreakerState {
    fn new(failure_threshold: usize, break_duration: Duration, recovery_threshold: usize) -> Self {
        Self {
            failure_threshold: failure_threshold.max(1),
            break_duration: if break_duration.is_zero() {
                Duration::from_secs(60)
            } else {
                break_duration
            },
            recovery_threshold: recovery_threshold.max(1),
            mode: CircuitBreakerMode::Closed,
            failure_count: 0,
            half_open_success_count: 0,
            open_until: None,
        }
    }

    fn try_allow_request(&mut self) -> Option<String> {
        if self.mode != CircuitBreakerMode::Open {
            return None;
        }

        let Some(open_until) = self.open_until else {
            self.mode = CircuitBreakerMode::HalfOpen;
            return None;
        };

        let now = Instant::now();
        if now >= open_until {
            self.mode = CircuitBreakerMode::HalfOpen;
            self.failure_count = 0;
            self.half_open_success_count = 0;
            self.open_until = None;
            return None;
        }

        let remaining = open_until.saturating_duration_since(now).as_secs().max(1);
        Some(format!("熔断开启，剩余约 {remaining} 秒"))
    }

    fn report_success(&mut self) {
        if self.mode == CircuitBreakerMode::HalfOpen {
            self.half_open_success_count += 1;
            if self.half_open_success_count >= self.recovery_threshold {
                self.reset_to_closed();
            }
            return;
        }

        self.failure_count = 0;
    }

    fn report_failure(&mut self) {
        if self.mode == CircuitBreakerMode::HalfOpen {
            self.open();
            return;
        }

        self.failure_count += 1;
        if self.failure_count >= self.failure_threshold {
            self.open();
        }
    }

    fn open(&mut self) {
        self.mode = CircuitBreakerMode::Open;
        self.failure_count = 0;
        self.half_open_success_count = 0;
        self.open_until = Some(Instant::now() + self.break_duration);
    }

    fn reset_to_closed(&mut self) {
        self.mode = CircuitBreakerMode::Closed;
        self.failure_count = 0;
        self.half_open_success_count = 0;
        self.open_until = None;
    }
}

fn build_full_url(base_url: &str, path_and_query: &str) -> String {
    let base = base_url.trim_end_matches('/');
    if path_and_query == "/" {
        return base.to_string();
    }
    format!("{base}{path_and_query}")
}

#[cfg(test)]
mod tests {
    use super::{CircuitBreakerMode, CircuitBreakerState, TransportEndpoint};
    use crate::options::{CentralServiceCircuitBreakerOptions, CentralServiceEndpointOptions};

    #[test]
    fn circuit_breaker_reopens_after_half_open_failure() {
        let mut state = CircuitBreakerState::new(1, std::time::Duration::from_millis(10), 1);
        state.report_failure();
        assert_eq!(Some("熔断开启，剩余约 1 秒".to_string()), state.try_allow_request());
        std::thread::sleep(std::time::Duration::from_millis(15));
        assert_eq!(None, state.try_allow_request());
        assert_eq!(CircuitBreakerMode::HalfOpen, state.mode);
        state.report_failure();
        assert_eq!(CircuitBreakerMode::Open, state.mode);
    }

    #[test]
    fn transport_endpoint_uses_default_attempts() {
        let mut endpoint = CentralServiceEndpointOptions::new("http://127.0.0.1:15700");
        endpoint.circuit_breaker = Some(CentralServiceCircuitBreakerOptions {
            failure_threshold: 1,
            break_duration_minutes: 1,
            recovery_threshold: 1,
        });
        let endpoint = endpoint.normalize(0).expect("normalize endpoint");
        let transport = TransportEndpoint::from_options(&endpoint);
        assert_eq!(2, transport.max_attempts);
    }
}
